src/service/position-observer-impl.js

```diff
@@ -0,0 +1,410 @@
+        entry.element./*OK*/getBoundingClientRect());
+    const win = this.ampdoc_.win;
+
+    let viewportWidth = win./*OK*/innerWidth;
+    let viewportHeight = win./*OK*/innerHeight;
+    if (!viewportWidth || !viewportHeight) {
+      const el = win.document.documentElement;
+      viewportWidth = el./*OK*/clientWidth;
+      viewportHeight = el./*OK*/clientHeight;
+    }
+    const viewportBox = layoutRectLtwh(0, 0, viewportWidth, viewportHeight);
+
+    // Return { positionRect: <LayoutRectDef>, viewportRect: <LayoutRectDef>}
+    entry.trigger(/** @type {PositionInViewportEntryDef}*/ ({
+      positionRect: elementBox,
+      viewportRect: viewportBox,
+    }));
+  }
+}
+
+export class InaboxPositionObserver extends AbstractPositionObserver {
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+    super(ampdoc);
+
+    /** @private {!AmpDocPositionObserver} */
+    this.positionObserver_ = new AmpDocPositionObserver(ampdoc);
+
+    // TODO(@zhouyx, #9208) support fidelity
+    this.effectiveFidelity_ = PositionObserverFidelity.LOW;
+
+    /** @private {?PositionInViewportEntryDef} */
+    this.iframePosition_ = null;
+
+    /** @private {?function()} */
+    this.unlistenHost_ = null;
+  }
+
+  /** @override */
+  observe(element, fidelity, handler) {
+    super.observe(element, fidelity, handler);
+    this.positionObserver_.observe(element, fidelity, position => {
+      element.inIframePositionRect = position.positionRect;
+    });
+  }
+
+  /** @override */
+  unobserve(element) {
+    super.unobserve(element);
+    this.positionObserver_.unobserve(element);
+  }
+
+  /** @override */
+  startCallback() {
+    // TODO(@zhouyx, #9208) need to add support for AMP host
+    if (!this.ampdoc_.win.context) {
+      dev.error(TAG, 'no window context for inbox window');
+    }
+
+    // TODO(@zhouyx, #9208) Migrate to using ampContext
+    this.unlistenHost_ = this.ampdoc_.win.context.observePosition(
+      position => {
+        this.onMessageReceivedHandler_(
+            /** @type {PositionInViewportEntryDef} */ (position));
+      });
+  }
+
+  /** @override */
+  stopCallback() {
+    if (this.unlistenHost_) {
+      this.unlistenHost_();
+      this.unlistenHost_ = null;
+    }
+  }
+
+  /**
+   * @param {!PositionInViewportEntryDef} iframePosition
+   * @private
+   */
+  onMessageReceivedHandler_(iframePosition) {
+    // iframe position change. recalculate element position.
+    // Cache iframe position for later usage.
+    this.iframePosition_ = iframePosition;
+
+    // do this in vsyn.measure
+    this.vsync_.measure(() => {
+      this.pass_();
+    });
+  }
+
+  /** @private */
+  pass_() {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      this.updateEntryPosition_(entry);
+    }
+  }
+
+  /** @override */
+  updateEntryPosition_(entry) {
+    if (!this.iframePosition_ || !this.iframePosition_.positionRect) {
```

Comment:
Is this a race condition? I.e. if update arrives for a child, but iframePosition is not available yet, it will be skipped. When `this.positionObserver_.observe` is triggered, this is not revisited. Right?